### PR TITLE
Introduce ChapterDelta and USX roundtripping tests

### PR DESCRIPTION
**Pull request by marksvc**
- dev: fix method name UpdateParatextBookAsync
- dev: add to comment on Chapter.IsValid
- dev: clean up test GetParatextChaptersAsDeltas_BeginsWithExpectedContent
- test: add ToUsx_AdjacentChar_SpaceBetweenRetained
- test: DeltaUsxMapper, ParatextSyncRunner: add some roundtripping
- test: add Hebrew sample data
- test: DeltaUsxMapper roundtrips Hebrew Bible

I am okay if this is squashed together as one commit when merged.

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2314)

### CODE REVIEW

**Comment by marksvc** [comment]
This does not change behaviour, so testing is not required.

**Review by marksvc — COMMENTED** [review]
Tests are introduced (or added to) to show that we roundtrip data back and forth between ChapterDelta and USX successfully. This is significant for not unintentionally changing user data that comes from Paratext.

A Hebrew Bible is added. It is read, converted to USX as an XDocument, converted to ChapterDeltas, converted back to XDocument USX, and compared to the earlier XDocument USX.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-:-NpMID0J-4piv256_BT1:b-ionun4)* status: 0 of 74 files reviewed, all discussions resolved

___
*[`-- commits` line 3 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMIVtaDPUomN31xoni):*
Changes can be viewed as individual commits, or as a whole.
___
*[`src/SIL.XForge.Scripture/Services/ParatextSyncRunner.cs` line 1040 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMJV17DMrLGUg78Ap8:-NpMJV17DMrLGUg78Ap9:b-fat251) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/src/SIL.XForge.Scripture/Services/ParatextSyncRunner.cs#L1040)):*
> ```Smalltalk
>     }
> 
>     internal static XDocument UsxToXDocument(string bookUsx)
> ```

This is split out into its own method so I can re-use it in a unit test.
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 62 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMJexc9rqXPWTbTDhW:-NpMJexc9rqXPWTbTDhX:b-5bqv62) ([raw file](https://github.com/sillsdev/web-xforge/blob/87d4abffe1ff2ac75c03733a694a8c13566fa86e/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L62)):*
> ```Smalltalk
>                 .InsertText("Verse text.", "verse_1_1")
>                 .InsertPara("p")
>                 .Insert("\n")
> ```

These newlines are trimmed and not something that gets roundtripped. This is described below in test ToUsx_CollapsesAdjacentNewlines. I removed the `Insert.("\n")` here to show the roundtripping of the rest of the content.
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 73 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMK6T53wdpYsqpOLGX:-NpMK6T67_n3CPUkZdxD:bqhs73h) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L73)):*
> ```Smalltalk
>         Assert.IsTrue(XNode.DeepEquals(newUsxDoc, expected));
> 
>         // And we should be able to roundtrip it back.
> ```

For a few tests I added some code to show that we roundtrip the content successfully.
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 921 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMKgmw3GBB2ysfNmIJ:-NpMKgmw3GBB2ysfNmIK:b-fzu0ev) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L921)):*
> ```Smalltalk
>                 .Insert("\n")
>                 .Insert("\n")
>                 .Insert("\n")
> ```

These newlines are trimmed in DeltaUsxMapper. This test demonstrates that.
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 935 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMKkiu1ZXXuehEADeF:-NpMKkiu1ZXXuehEADeG:b-du8qxi) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L935)):*
> ```Smalltalk
>                 .InsertVerse("1")
>                 .InsertText("Verse text.", "verse_1_1")
>                 .InsertPara("p")
> ```

The test demonstrates that not specifying any newlines here results in the same thing as specifying several.
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3412 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMKxIH1-QzZQ493Pyt:-NpMKxIH1-QzZQ493Pyu:b2x9gpq) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3412)):*
> ```Smalltalk
> 
>     [Test]
>     public void RoundTrip_Hebrew() => RoundTripTestHelper("hebrew");
> ```

The Hebrew Bible sample data can be roundtripped.
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3417 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpML4Ba6e3CbnOuuBOP:-NpML4Ba6e3CbnOuuBOQ:b6uhpwv) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3417)):*
> ```Smalltalk
>     {
>         DirectoryInfo codeRepoRoot = Directory.GetParent(Directory.GetCurrentDirectory()).Parent.Parent.Parent.Parent;
>         var usfmFiles = new DirectoryInfo(Path.Combine(codeRepoRoot.FullName, "samples", project))
> ```

Collect all the Scripture book .usfm files.
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3442 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMLHdD32jNMURulVGJ:-NpMLHdD32jNMURulVGK:b8tqqph) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3442)):*
> ```Smalltalk
>             using XmlNodeReader reader = new(bookUsxLoading);
>             reader.MoveToContent();
>             XDocument bookUsx = XDocument.Load(reader);
> ```

Get an XDocument from the USFM.
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3451 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMLLvJ5jt0PEaFpb9J:-NpMLLvK6IM6FDTJWj1f:b-bqkrl5) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3451)):*
> ```Smalltalk
> 
>             // SUT part 1
>             List<ChapterDelta> chapterDeltas = mapper.ToChapterDeltas(bookUsx).ToList();
> ```

Convert the XDocument USX to ChapterDeltas.
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3459 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMLWKD-8wT-tZSEQ6u:-NpMLWKD-8wT-tZSEQ6v:b-sfll3m) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3459)):*
> ```Smalltalk
> 
>             // SUT part 2
>             XDocument roundTrippedUsx = mapper.ToUsx(
> ```

Convert the ChapterDeltas back to XDocument USX, and compare.
___
*[`test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs` line 3014 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMLjnaA5l6GGB0kLEr:-NpMLjnaA5l6GGB0kLEs:bckdb4j) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs#L3014)):*
> ```Smalltalk
> 
>     [Test]
>     public async Task CoreRoundtrip_NoUnexpectedDataChanges()
> ```

ParatextSyncRunner uses DeltaUsxMapper to convert between XDocument USX and ChapterDeltas. This conversion can be tested in DeltaUsxMapperTests. But ParatextSyncRunner does some extra pieces in addition to this. This integration test runs through a couple core methods and shows them roundtripping some data.
___
*[`test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs` line 3170 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpMMIwi-PbTHzOOC8ji:-NpMMIwi-PbTHzOOC8jj:b-rcj52y) ([raw file](https://github.com/sillsdev/web-xforge/blob/12afe7991b93ea608b0b1b0e3a00e70c117f0968/test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs#L3170)):*
> ```Smalltalk
>         sb2.AppendLine("  <para style=\"p\">");
>         sb2.AppendLine("    <verse number=\"1\" style=\"v\" />");
>         sb2.AppendLine($"{newText}<char style=\"w\">pass</char> <char style=\"w\">in</char> the days</para>");
> ```

(This line is different than the earlier USX, with `{newText}`.)

**Review by pmachapman — CHANGES_REQUESTED** [review]
> Reviewed 1 of 74 files at r1, 73 of 73 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2314)* status: all files reviewed, 3 unresolved discussions (waiting on @marksvc)

___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3465 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpRwgbo5QJZjqxtE-06:-NpRwgbo5QJZjqxtE-07:becahgn) ([raw file](https://github.com/sillsdev/web-xforge/blob/8d8d116a39a1c05c7a68bb7da3e92b11ea00ed88/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3465)):*
> ```Smalltalk
>             Assert.IsTrue(XNode.DeepEquals(roundTrippedUsx, bookUsx), $"Trouble in {inputFile.bookPath}");
>         }
>     }
> ```

We should have data used by unit tests in the same place as the unit tests to help with perspicuity, so I thought maybe we could have the unit test just read the data from a zip file?

This does slow the test down to take about a second on my computer, so if you would rather the text file approach, that is cool. My only suggestion in that case is to use the Assembly location rather than the current directory to get your starting point to find the samples directory.

If you do want to use a zip, create a directory called `SampleData` in the `test/SIL.XForge.Scripture.Tests` directory, then add a zip file called [heb_usfm.zip](https://ebible.org/Scriptures/heb_usfm.zip) containing the USFM:



We could change this test to be:

```cs
using System.IO.Compression;
using System.Reflection;
using System.Threading.Tasks;


[Test]
public async Task RoundTrip_Hebrew() => await RoundTripTestHelper("heb_usfm");

private async Task RoundTripTestHelper(string project)
{
    string zipFilePath = Path.Combine(GetPathToTestProject(), "SampleData", $"{project}.zip");
    await using FileStream zipFileStream = new FileStream(zipFilePath, FileMode.Open, FileAccess.Read);
    using ZipArchive archive = new ZipArchive(zipFileStream, ZipArchiveMode.Read);
    Assert.That(archive.Entries.Any(), "setup. unexpected input size.");
    foreach (ZipArchiveEntry entry in archive.Entries)
    {
        string bookCode = Regex.Match(entry.Name, @".*-([A-Z0-9]{3}).*\.usfm").Groups[1].Value;
        if (entry.Name.EndsWith(".usfm", StringComparison.OrdinalIgnoreCase) && bookCode is not ("FRT" or "INT"))
        {
            await using Stream entryStream = entry.Open();
            using StreamReader reader = new StreamReader(entryStream);

            // Read and stream the contents of the text file
            string bookUsfm = await reader.ReadToEndAsync();
            XmlDocument bookUsxLoading = Paratext.Data.UsfmToUsx.ConvertToXmlDocument(
                new Paratext.Data.MockScrStylesheet("usfm.sty"),
                bookUsfm
            );
            using XmlNodeReader nodeReader = new(bookUsxLoading);
            nodeReader.MoveToContent();
            XDocument bookUsx = XDocument.Load(nodeReader);
            // Record the usx version string to make it match when later compared.
            string usxVersion = bookUsx.Elements("usx").First().Attribute("version")!.Value;
            // Record any text in the book node, which some books have, like <book code="GEN">- American Standard
            // Version</book>
            string? bookDesc = bookUsx.Elements("usx").Elements("book").First().FirstNode?.ToString();
            DeltaUsxMapper mapper = new(_mapperGuidService, _logger, _exceptionHandler);

            // SUT part 1
            List<ChapterDelta> chapterDeltas = mapper.ToChapterDeltas(bookUsx).ToList();

            IEnumerable<XElement> chaptersToProcess = bookUsx
                .Elements("usx")
                .Elements("chapter")
                .Select(x => Chapter(x.Attribute("number")!.Value));

            // SUT part 2
            XDocument roundTrippedUsx = mapper.ToUsx(
                Usx(bookCode, bookDesc, usxVersion, chaptersToProcess),
                chapterDeltas
            );
            Assert.IsTrue(XNode.DeepEquals(roundTrippedUsx, bookUsx), $"Trouble in {entry.Name}");
        }
    }
}

private static string GetPathToTestProject()
{
    string pathAssembly = Assembly.GetExecutingAssembly().Location;
    string folderAssembly = Path.GetDirectoryName(pathAssembly) ?? string.Empty;
    if (!folderAssembly.EndsWith(Path.DirectorySeparatorChar)) folderAssembly += Path.DirectorySeparatorChar;
    return Path.GetFullPath($"{folderAssembly}..{Path.DirectorySeparatorChar}..{Path.DirectorySeparatorChar}..{Path.DirectorySeparatorChar}");
}
```

Also, it is better to get the path from the assembly location, not the current directory, as the current directory may not be as reliable.



_Code quote:_
```Smalltalk
    [Test]
    public void RoundTrip_Hebrew() => RoundTripTestHelper("hebrew");

    public void RoundTripTestHelper(string project)
    {
        DirectoryInfo codeRepoRoot = Directory.GetParent(Directory.GetCurrentDirectory()).Parent.Parent.Parent.Parent;
        var usfmFiles = new DirectoryInfo(Path.Combine(codeRepoRoot.FullName, "samples", project))
            .GetFiles("*.usfm")
            .OrderBy((FileInfo file) => file.FullName, StringComparer.Ordinal)
            .Select(
                file =>
                    new
                    {
                        bookPath = file.FullName,
                        bookCode = Regex.Match(file.FullName, @".*-([A-Z0-9]{3}).*\.usfm").Groups[1].Value
                    }
            )
            // Exclude peripheral books that are not part of the usx-sf.rnc schema.
            .Where(file => file.bookCode != "FRT" && file.bookCode != "INT");

        Assert.That(usfmFiles.Count() > 0, "setup. unexpected input size.");

        foreach (var inputFile in usfmFiles)
        {
            string bookUsfm = File.ReadAllText(inputFile.bookPath);
            XmlDocument bookUsxLoading = Paratext.Data.UsfmToUsx.ConvertToXmlDocument(
                new Paratext.Data.MockScrStylesheet("usfm.sty"),
                bookUsfm
            );
            using XmlNodeReader reader = new(bookUsxLoading);
            reader.MoveToContent();
            XDocument bookUsx = XDocument.Load(reader);
            // Record the usx version string to make it match when later compared.
            string usxVersion = bookUsx.Elements("usx").First().Attribute("version").Value;
            // Record any text in the book node, which some books have, like <book code="GEN">- American Standard
            // Version</book>
            string? bookDesc = bookUsx.Elements("usx").Elements("book").First().FirstNode?.ToString();
            DeltaUsxMapper mapper = new(_mapperGuidService, _logger, _exceptionHandler);

            // SUT part 1
            List<ChapterDelta> chapterDeltas = mapper.ToChapterDeltas(bookUsx).ToList();

            IEnumerable<XElement> chaptersToProcess = bookUsx
                .Elements("usx")
                .Elements("chapter")
                .Select(x => Chapter(x.Attribute("number").Value));

            // SUT part 2
            XDocument roundTrippedUsx = mapper.ToUsx(
                Usx(inputFile.bookCode, bookDesc, usxVersion, chaptersToProcess),
                chapterDeltas
            );
            Assert.IsTrue(XNode.DeepEquals(roundTrippedUsx, bookUsx), $"Trouble in {inputFile.bookPath}");
        }
    }
```
___
*[`test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs` line 3025 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpS0cKP0R-wFQoKdtuj:-NpS0cKP0R-wFQoKdtuk:b-1extda) ([raw file](https://github.com/sillsdev/web-xforge/blob/8d8d116a39a1c05c7a68bb7da3e92b11ea00ed88/test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs#L3025)):*
> ```Smalltalk
>         mapperGuidService = new TestGuidService();
>         logger = Substitute.For<ILogger<DeltaUsxMapper>>();
>         exceptionHandler = Substitute.For<IExceptionHandler>();
> ```

We should either have these in the TestEnvironment, or change it to:
```cs
        IGuidService mapperGuidService = new TestGuidService();
        ILogger<DeltaUsxMapper> logger = Substitute.For<ILogger<DeltaUsxMapper>>();
        IExceptionHandler exceptionHandler = Substitute.For<IExceptionHandler>();
```



_Code quote:_
```Smalltalk
        IGuidService mapperGuidService;
        ILogger<DeltaUsxMapper> logger;
        IExceptionHandler exceptionHandler;
        mapperGuidService = new TestGuidService();
        logger = Substitute.For<ILogger<DeltaUsxMapper>>();
        exceptionHandler = Substitute.For<IExceptionHandler>();
```
___
*[`test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs` line 3075 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpS0uvYE1z9tJMjcGS1:-NpS0uvYE1z9tJMjcGS2:b-r9ac0t) ([raw file](https://github.com/sillsdev/web-xforge/blob/8d8d116a39a1c05c7a68bb7da3e92b11ea00ed88/test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs#L3075)):*
> ```Smalltalk
>                     DataInSync = true
>                 },
>                 ParatextUsers = new List<ParatextUserProfile>
> ```

nit You can remove the ` new List<ParatextUserProfile>`



_Code quote:_
```Smalltalk
ParatextUsers = new List<ParatextUserProfile>
```

**Comment by Nateowami** [comment]
Where did the test data come from? What is the license on it? Do we have a README or something similar in the same directory as the test data to explain a bit about it?

**Comment by pmachapman** [comment]
> Where did the test data come from? What is the license on it?

The data is Public Domain, and from https://ebible.org/find/details.php?id=heb. There is a copyr.htm file in the sample directory with copyright info, but I admit it took me a couple of minutes to figure out it was from ebible.org.

**Comment by Nateowami** [comment]
:facepalm: Clearly I need to start reading.

**Comment by pmachapman** [comment]
> 🤦 Clearly I need to start reading.

Its OK I had the same thoughts rolling round my head and took a few minutes to figure out. I was searching for strings from the USFM file in Google until I realized copyr.htm was the copyright file!

**Review by marksvc — COMMENTED** [review]
I agree that the directory and copr.htm file takes some looking at to figure out that it's from eBible.org. Maybe in part because I usually view the copr.htm file in a text editor and need to look between all the html to find the content :)
I think it's figure-out-able enough and so won't add another README to further describe where it's from unless you guys think I should.
I added some more information about license and source into the git commit message, FWIW.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-:-NpVnxiq9-cJz5XqjzoV:b-wtnyzg)* status: all files reviewed, 2 unresolved discussions (waiting on @pmachapman)

___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3465 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpRwgbo5QJZjqxtE-06:-NpVca0G0kJTx0GRACy0:b-r5o5e8) ([raw file](https://github.com/sillsdev/web-xforge/blob/8d8d116a39a1c05c7a68bb7da3e92b11ea00ed88/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3465)):*
> await using FileStream zipFileStream \= new FileStream(zipFilePath, FileMode.Open, FileAccess.Read);

How did you know you could use `await` here? Looking at MSDN I didn't understand something that would have led me use `await`. 

Hmm, same question for `await using Stream entryStream \= entry.Open();`.

Good point, the current directory might be easily something different :)

I am using the zip file. Thank you for the code along with your suggestion!
In my own tests I don't see that on average it is any slower to run the test, when using the zip.
___
*[`test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs` line 3025 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpS0cKP0R-wFQoKdtuj:-NpVnGsv2JXhjJLm7xaV:bl3mrru) ([raw file](https://github.com/sillsdev/web-xforge/blob/8d8d116a39a1c05c7a68bb7da3e92b11ea00ed88/test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs#L3025)):*
<details><summary><i>Previously, pmachapman (Peter Chapman) wrote…</i></summary><blockquote>

We should either have these in the TestEnvironment, or change it to:
```cs
        IGuidService mapperGuidService = new TestGuidService();
        ILogger<DeltaUsxMapper> logger = Substitute.For<ILogger<DeltaUsxMapper>>();
        IExceptionHandler exceptionHandler = Substitute.For<IExceptionHandler>();
```
</blockquote></details>

Yes, thank you :). I kept them here since they are part of using a real DeltaUsxMapper rather than a substitute DeltaUsxMapper.
___
*[`test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs` line 3075 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpS0uvYE1z9tJMjcGS1:-NpVnpOg-5ENfmRg6R1C:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/8d8d116a39a1c05c7a68bb7da3e92b11ea00ed88/test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs#L3075)):*
<details><summary><i>Previously, pmachapman (Peter Chapman) wrote…</i></summary><blockquote>

nit You can remove the ` new List<ParatextUserProfile>`
</blockquote></details>

Done.

**Review by pmachapman — APPROVED** [review]
:lgtm:

> Reviewed 73 of 73 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-:-NpVsHY-8B8mmuXiudS7:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @marksvc)

___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3465 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpRwgbo5QJZjqxtE-06:-NpVqvox2jzvwQ-4PeqO:b103ima) ([raw file](https://github.com/sillsdev/web-xforge/blob/8d8d116a39a1c05c7a68bb7da3e92b11ea00ed88/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3465)):*
> How did you know you could use `await` here? Looking at MSDN I didn't understand something that would have led me use `await`.

Resharper told me. Also, Stream is `IAsyncDisposable` (that is all that await is referring to - the disposal). I find that the only Streams that do not play well with async are XML Stream, which you have to set the Async property manually on them to make them asynchronous.

> In my own tests I don't see that on average it is any slower to run the test, when using the zip.

Cool!
___
*[`test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs` line 3023 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpVsMyEEEqUkXWzmvvA:-NpVsMyEEEqUkXWzmvvB:bja5nc3) ([raw file](https://github.com/sillsdev/web-xforge/blob/2ac4956cf282f4eafa8dfaebb6fab7813f367c61/test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs#L3023)):*
> ```Smalltalk
> 
>         // Not using something like SetupSFData() because DeltaUsxMapper is not a substitute.
>         Book[] books = [new Book("RUT", 8)];
> ```

Ooo the new C# 12 syntax for Collection Expressions.

I have been too scared to use this until the .NET 8.0 update, so I am glad to see all the tests passed with this.



_Code quote:_
```Smalltalk
[new Book("RUT", 8)];
```

**Review by marksvc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2314)* status: all files reviewed, 2 unresolved discussions (waiting on @github-advanced-security[bot] and @pmachapman)

___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3465 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpRwgbo5QJZjqxtE-06:-NpW4VdWBWSeYiYQHSMF:buffuhg) ([raw file](https://github.com/sillsdev/web-xforge/blob/8d8d116a39a1c05c7a68bb7da3e92b11ea00ed88/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3465)):*
> Stream is `IAsyncDisposable` (that is all that await is referring to - the disposal). I

Ahh. Right. `await using ...`, not `using Stream entryStream = await entry.Open();`
___
*[`test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs` line 3483 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/2314#gh-1473312840:-NpW54oh-ZlUvLvDabm0:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/2ac4956cf282f4eafa8dfaebb6fab7813f367c61/test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs#L3483)):*
<details><summary><i>Previously, github-advanced-security[bot] wrote…</i></summary><blockquote>

## Missed ternary opportunity

Both branches of this 'if' statement return - consider using '?' to express intent better.

[Show more details](https://github.com/sillsdev/web-xforge/security/code-scanning/714)
</blockquote></details>

Done.
___
*[`test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs` line 3023 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/2314#-NpVsMyEEEqUkXWzmvvA:-NpW5CTZCcyqUQ6HFv2C:b-ef17s2) ([raw file](https://github.com/sillsdev/web-xforge/blob/2ac4956cf282f4eafa8dfaebb6fab7813f367c61/test/SIL.XForge.Scripture.Tests/Services/ParatextSyncRunnerTests.cs#L3023)):*
<details><summary><i>Previously, pmachapman (Peter Chapman) wrote…</i></summary><blockquote>

Ooo the new C# 12 syntax for Collection Expressions.

I have been too scared to use this until the .NET 8.0 update, so I am glad to see all the tests passed with this.
</blockquote></details>

Yeah, VSCode has been nudging me about it for a bit :)

File: test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs [hunk 1]
```diff
@@ -3314,11 +3411,77 @@
...
+    private static XElement Book(string code, string? innerText = null)
+    {
+        if (innerText == null)
+        {
+            return new XElement("book", new XAttribute("code", code), new XAttribute("style", "id"));
+        }
+        else
+        {
+            return new XElement("book", new XAttribute("code", code), new XAttribute("style", "id"), innerText);
+        }
```
[the remark this answers is not in the document]
Comment (marksvc) [reply test/SIL.XForge.Scripture.Tests/Services/DeltaUsxMapperTests.cs]: Done.

**Comment by marksvc** [comment]
> marksvc requested a review from pmachapman

Oh, that's interesting. That must be automatic. I did change some code, but it was pretty slight. You don't need to have a look at it again, unless you beat the CI run :)

**Review by pmachapman — APPROVED** [review]
> Reviewed 72 of 72 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2314)* status: all files reviewed, 1 unresolved discussion (waiting on @github-advanced-security[bot])

**Review by pmachapman — APPROVED** [review]
> Dismissed @github-advanced-security[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/2314)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @marksvc)

**Pull request merged**